pinot-controller/src/main/java/org/apache/pinot/controller/api/upload/ZKOperator.java

```diff
@@ -214,14 +216,34 @@ private void processNewSegment(SegmentMetadata segmentMetadata, URI finalSegment
             .info("Moved segment {} from temp location {} to {}", segmentName, currentSegmentLocation.getAbsolutePath(),
                 finalSegmentLocationURI.getPath());
       } catch (Exception e) {
-        LOGGER.error("Could not move segment {} from table {} to permanent directory", segmentName, tableNameWithType, e);
+        LOGGER
+            .error("Could not move segment {} from table {} to permanent directory", segmentName, tableNameWithType, e);
         throw new RuntimeException(e);
       }
     } else {
       LOGGER.info("Skipping segment move, keeping segment {} from table {} at {}", segmentName, tableNameWithType,
           zkDownloadURI);
     }
     _pinotHelixResourceManager.addNewSegment(tableNameWithType, segmentMetadata, zkDownloadURI, crypter);
+
+    // Update zk metadata customer map for offline table
+    String segmentZKMetadataCustomMapModifierStr =
+        headers.getHeaderString(FileUploadDownloadClient.CustomHeaders.SEGMENT_ZK_METADATA_CUSTOM_MAP_MODIFIER);
+    if (TableType.OFFLINE == TableNameBuilder.getTableTypeFromTableName(tableNameWithType)
```

Comment:
Is this check only for reading the correct type of the ZK metadata? If so, let's wait for #7255 to be merged first, then this check can be avoided.
This metadata modifier should apply to any segment instead of only segments from OFFLINE table

Reply:
Yes, we need to use `OfflineSegmentZKMetadata` here because currently we cant initialize `SegmentZKMetadata`. #7255 should fix this.